File: artifacts/api-server/src/routes/messages.ts
import { Router, type IRouter } from "express";
import { eq } from "drizzle-orm";
import { RateMessageBody } from "@workspace/api-zod";
import { db, messagesTable, conversationsTable } from "@workspace/db";
import { requireAuth } from "../middlewares/auth";

const router: IRouter = Router();

router.post("/messages/:id/feedback", requireAuth, async (req, res) => {
  const id = Number(req.params.id);
  const body = RateMessageBody.parse(req.body);
  const [msg] = await db.select().from(messagesTable).where(eq(messagesTable.id, id));
  if (!msg) {
    res.status(404).json({ error: "Message not found", status: 404 });
    return;
  }
  const [convo] = await db
    .select()
    .from(conversationsTable)
    .where(eq(conversationsTable.id, msg.conversationId));
  if (!convo || convo.userId !== req.user!.userId) {
    res.status(403).json({ error: "Forbidden", status: 403 });
    return;
  }
  const [updated] = await db
    .update(messagesTable)
    .set({ rating: body.rating, feedbackComment: body.comment ?? null })
    .where(eq(messagesTable.id, id))
    .returning();
  res.json({
    id: updated.id,
    conversationId: updated.conversationId,
    role: updated.role,
    content: updated.content,
    citations: updated.citations,
    canAnswer: updated.canAnswer,
    latencyMs: updated.latencyMs,
    rating: updated.rating,
    createdAt: updated.createdAt.toISOString(),
  });
});

export default router;

File: artifacts/api-server/app/routes/messages.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException

from app.auth import AuthedUser, require_auth
from app.db import get_db
from app.schemas import RateMessageBody
from app.serialize import serialize_message

router = APIRouter()


@router.post("/messages/{id}/feedback")
async def rate_message(
    id: str,
    body: RateMessageBody,
    user: AuthedUser = Depends(require_auth),
) -> dict[str, object]:
    try:
        msg_id = int(id)
    except (TypeError, ValueError):
        raise HTTPException(status_code=400, detail="Invalid id")
    if msg_id <= 0:
        raise HTTPException(status_code=400, detail="Invalid id")

    db = await get_db()
    msg = await db.messages.find_one({"_id": msg_id})
    if not msg:
        raise HTTPException(status_code=404, detail="Message not found")
    convo = await db.conversations.find_one({"_id": msg["conversationId"]})
    if not convo or convo.get("userId") != user.userId:
        raise HTTPException(status_code=403, detail="Forbidden")

    r = await db.messages.find_one_and_update(
        {"_id": msg_id},
        {"$set": {"rating": body.rating, "feedbackComment": body.comment}},
        return_document=True,
    )
    if not r:
        raise HTTPException(status_code=404, detail="Message not found")
    return serialize_message(r)
